ReactWindows/ReactNative.Shared/UIManager/UIManagerModule.cs

```diff
@@ -111,33 +111,45 @@ public UIImplementation UIImplementation
         /// JavaScript can use the returned tag with to add or remove children 
         /// to this view through <see cref="manageChildren(int, int[], int[], int[], int[], int[])"/>.
         /// </remarks>
-        public int AddMeasuredRootView(SizeMonitoringCanvas rootView)
+        public int AddMeasuredRootView(ReactRootView rootView)
         {
+            // Called on main dispatcher thread
+            DispatcherHelpers.AssertOnDispatcher();
+
             var tag = _nextRootTag;
             _nextRootTag += RootViewTagIncrement;
 
-            var width = rootView.ActualWidth;
-            var height = rootView.ActualHeight;
-
             var context = new ThemedReactContext(Context);
-            _uiImplementation.RegisterRootView(rootView, tag, width, height, context);
 
-            var resizeCount = 0;
-            rootView.SetOnSizeChangedListener((sender, args) =>
+            DispatcherHelpers.RunOnDispatcher(rootView.Dispatcher, () =>
             {
-                var currentCount = ++resizeCount;
-                var newWidth = args.NewSize.Width;
-                var newHeight = args.NewSize.Height;
+                var width = rootView.ActualWidth;
+                var height = rootView.ActualHeight;
 
                 _layoutActionQueue.Dispatch(() =>
                 {
```

Comment:
My concern here is the `RegisterRootView` is being dispatched from an asynchronous call. Do we have guarantees that the `DispatcherHelpers.RunOnDispatcher` call will evaluate before the JS runtime starts pumping out `UIManagerModule` method calls for that root view?

Reply:
I think lots of these matters depend on the order work items are posted to a queue (and of course on not jumping the gun in some cases). 
From what I understand this _layoutActionQueue.Dispatch is the first work item pertaining for this root view and its children since the JS side doesn't even know about the activation of the view yet.
The caller of AddMeasuredRootView (i.e. AttachMeasuredRootViewToInstance) then continues with running the JS "application". New views added/updated can only happen through dispatching new items to that layoutActionQueue.

I know there are some items bypassing the layout manager queue, like animations. But they are either UI agnostic (like animated values, etc.) OR they affect views that have to have been created in the first place.
So I think we're safe.



---
In reply to: [165940343](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r165940343) [](ancestors = 165940343)